feat(appmesh) : implementing mutual TLS  feature

Seiya6329 opened the pull request:
**Note:**  
- this is for internal-review purpose only. I will submit the PR to main repo once I receive green from the team.
- This review **DOES NOT** include changes on `README` file yet. `README` file needs to be updated to include mTLS implementation example.

----
This is the last part of the series to implement mTLS feature and closes #12733.
For the breakdown on series, please refer to
[this comment](https://github.com/aws/aws-cdk/pull/14782#discussion_r636334863).

#### Collaborators
@alexbrjo and @dfezzie. Thank you!

#### REV
 - Adding SDS certificate source to `TlsCertificate` and `TlsValidationTrust`.
 - Adding `subjectAlternativeNames` property to `TlsValidation`.
 - Adding `certificate` property to `TlsClientPolicy`
 - Adding `validation` property to `TlsListener`

#### Design Note:
 - Client certificates in a TLS Client Policy and server validation in a listener TLS configuration can only be sourced from `SDS` or `File` certificate. ACM certificate is currently not supported ([ref](https://docs.aws.amazon.com/app-mesh/latest/userguide/mutual-tls.html))
   - when ACM certificate is selected, run-time error is thrown.

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

Seiya6329 commented on packages/@aws-cdk/aws-appmesh/lib/tls-validation.ts:
```diff
@@ -62,12 +94,19 @@ export abstract class TlsValidationTrust {
   }
 
+  /**
+   * TLS Validation Context Trust for Enovoy' service discovery service.
```
Envoy is misspelled. Will address this.

Seiya6329 replied:
Addressed.

alexbrjo commented on packages/@aws-cdk/aws-appmesh/lib/tls-client-policy.ts:
```diff
@@ -1,9 +1,18 @@
+   * Represents a client's TLS certificate.
+   * Define this to enable mutual TLS authentication.
+   *
+   * @default - no TLS certificate
```
nit: `@default - no client TLS Certificate`

Seiya6329 replied:
Addressed.

alexbrjo commented on packages/@aws-cdk/aws-appmesh/lib/tls-validation.ts:
```diff
@@ -125,11 +164,110 @@ class TlsValidationFileTrust extends TlsValidationTrust {
+
+  public bind(_scope: Construct): TlsValidationTrustConfig {
+    return {
+      virtualNodeClientTlsValidationTrust: {
+        sds: {
+          secretName: this.secretName,
+        },
+      },
+      virtualNodeListenerTlsValidationTrust: {
+        sds: {
+          secretName: this.secretName,
+        },
+      },
+      virtualGatewayClientTlsValidationTrust: {
+        sds: {
+          secretName: this.secretName,
+        },
+      },
+      virtualGatewayListenerTlsValidationTrust: {
+        sds: {
+          secretName: this.secretName,
+        },
+      },
```
You can reduce the repetition (mentioned this on the last PR I think). Example: https://github.com/alexbrjo/aws-cdk/commit/aea876e1d62fb34437a774c9861bdcd9c6b91527

Seiya6329 replied:
Reduced the repetition to simplify the code.

alexbrjo commented on packages/@aws-cdk/aws-appmesh/lib/tls-validation.ts:
```diff
@@ -125,11 +164,110 @@ class TlsValidationFileTrust extends TlsValidationTrust {
+  /**
+   * VirtualNode CFN configuration for subject alternative names secured by the certificate.
+   */
+  readonly virtualNodeSans: CfnVirtualGateway.SubjectAlternativeNameMatchersProperty;
```
nit: you're not exposing it to the customer anywhere so this just a style thing but I like `virtualNodeSubjectAlternativeNames` instead of `virtualNodeSans`. It's could be confusing that some variable names have it spelled out and some have it abbreviated.

Seiya6329 replied:
Well noted. I switched the abbreviated to full name to match the style.

alexbrjo commented on packages/@aws-cdk/aws-appmesh/lib/tls-validation.ts:
```diff
@@ -125,11 +164,110 @@ class TlsValidationFileTrust extends TlsValidationTrust {
+   *
+   * @param subjectAlternativeNames The exact values to test against.
+   */
+  public static valuesAre(subjectAlternativeNames: string[]): SubjectiveAlternativeNamesMatch {
```
IMO we should rename this to `exact` because it's more explicit.

```
subjectAlternativeNames: appmesh.SubjectiveAlternativeNamesMatch.exactMatch(['san1', 'san2']),
```

Another consideration: Currently there is only one type of match (exact) but in the future we could have more. It's possible we'll allow users to define multiple match modes which would mean we'd have to break this part of the CDK API when we update it.

Not using a factory here could be more safe. Like this:

```
subjectAlternativeNames:
    exactMatch: []
    prefixMatch: [] # future?
```

 Thoughts? + @dfezzie

dfezzie replied:
+1 to your suggestions. I prefer `exact` over `valuesAre`

As long as it's semantically valid to have multiple different matches, that seems right.

Seiya6329 replied:
Good point AJ. I agree with you avoiding to use the factory method is safe option at the moment.

Seiya6329 commented:
REV2:
- Reducing the repetitions on `TlsValidationTrust` property since all of them are guaranteed to have same property.
- Addressed nits

Seiya6329 commented:
REV2:
- Reducing the repetitions on `TlsValidationTrust` property since all of them are guaranteed to have same property.
- Changing the `SubjectAlternativeNames` factory method from `valuesAre` to `exactMatch`.
- Addressed nits

Seiya6329 commented:
REV2:
- Reducing the repetitions on `TlsValidationTrust` property since all of them are guaranteed to have same property.
- Converting the `SubjectAlternativeNames` to interface to avoid using the factory method.
- Addressed nits

Seiya6329 commented:
**REV3:**
- Updating `README` file to include:
  - mTLS example
  - SDS certificate option under [Adding TLS to a listener](https://github.com/aws/aws-cdk/blob/master/packages/%40aws-cdk/aws-appmesh/README.md#adding-tls-to-a-listener) section

**NOTE:**
- Subject Alternative Names change is **NOT* explicitly called out in the `README` since I am thinking it is not a major feature. However, I did include `subjectAlternativeNames` property inside the mTLS example.

alexbrjo commented on packages/@aws-cdk/aws-appmesh/README.md:
```diff
@@ -277,6 +280,62 @@ const gateway = new appmesh.VirtualGateway(this, 'gateway', {
+  })],
+});
+
+// Provide a SDS client certificate when a server requests it and enable mutual TLS authentication.
```
this ARN is for an ACM cert; not SDS

Seiya6329 replied:
The ARN is actually used by the ACM cert but I understand the way I placed the comment makes it confusing.

I will move the comment.

alexbrjo commented on packages/@aws-cdk/aws-appmesh/README.md:
```diff
@@ -277,6 +280,62 @@ const gateway = new appmesh.VirtualGateway(this, 'gateway', {
+
+const node2 = new appmesh.VirtualNode(stack, 'node2', {
+  mesh,
+  serviceDiscovery: appmesh.ServiceDiscovery.dns('node'),
```
not important since this is an example but different Virtual Nodes should have different DNS names

Seiya6329 replied:
Thanks for pointing this out! It is helpful to know these specific details.

alexbrjo reviewed (approved):
Looks good!
